### ISSUE TITLE
Setting climate operation_mode often fails

### ISSUE THREAD

**Issue by mlowijs** [zabuldon/teslajsonpy#16]
Often when calling `set_operation_mode` on a Tesla climate entity, the call either fails with an error or does not do anything at all.

Would it be possible/sensible to add a vehicle wake up call to the `Climate.set_status` call?

**Comment by zabuldon** [comment]
@mlowijs Thank you for the report. I'll look into it ASAP.

**Comment by alandtse** [comment]
We probably need a wake up decorator or for loop to ensure the car has actually woken up before issuing the command.

**Comment by zabuldon** [comment]
@alandtse the problem is there is no "real" way to check if car woke up. I'll check the API methods again, but there is not normal documentation available (only outdated one).

**Comment by alandtse** [comment]
@zabuldon Can't we check the api is returning some car state? Or is it cached even if not woken up?

**Comment by zabuldon** [comment]
@alandtse I haven't seen something like that. But will re-check with a new application

**Comment by bart-roos** [comment]
As I described in https://github.com/home-assistant/home-assistant/issues/17272#issuecomment-435712406, I think the real issue is that the teslajsonpy library issues the wake_up and other commands in a fire-and-forget way. What it should do is check the actual result of these calls, and retry a couple of times when an empty result is returned. An empty result seems to indicate that the car didn't fully woke up yet. After a couple of tries, an actual result will be returned, indicating that the car woke up and the command is executed.

**Comment by zabuldon** [comment]
@bart-roos thanks you for details. I think I've found REAL cause of the issue. It is connected with your description.

**Comment by zabuldon** [comment]
@alandtse I've committed a "workaround" to fix this. Do you have time to test it?

**Comment by alandtse** [comment]
@zabuldon I haven't run into the problem consistently to be able to test this fix. However, reviewing the code I don't see where we increment the retry counter so I think there may be an infinite loop.

@bart-roos assuming we resolve the retry counter, do you see this issue consistently enough to test it?

**Comment by zabuldon** [comment]
@alandtse Thanks, just fixed that.

**Comment by bart-roos** [comment]
@alandtse, I've just added these fixes to my homeassistant setup and will test it during the next few days. in my case the original issue was consistent enough to be able to determine if the modifications are effective.

**Comment by alandtse** [comment]
Thanks @bart-roos, please make sure you use the [latest staging branch](https://github.com/zabuldon/teslajsonpy/compare/staging) and **not** the [commit ](https://github.com/zabuldon/teslajsonpy/commit/7b5ddc51b14da341a7a34220ec2d4a56384cbdb9)mentioned above.

**Comment by bart-roos** [comment]
The issue is not yet resolved, but I am getting these errors in the homeassistant logging:

```
2019-02-10 15:46:34 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "/opt/homeassistant/lib/python3.5/site-packages/homeassistant/helpers/service.py", line 289, in _handle_service_platform_call
    await getattr(entity, func)(**data)
  File "/usr/lib/python3.5/asyncio/futures.py", line 380, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 304, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 293, in result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/homeassistant/lib/python3.5/site-packages/homeassistant/components/climate/tesla.py", line 100, in set_operation_mode
    self.tesla_device.set_status(True)
  File "/opt/homeassistant/lib/python3.5/site-packages/teslajsonpy/Climate.py", line 68, in set_status
    if data['response']['result']:
TypeError: 'bool' object is not subscriptable
```

It seems that the climate setting is issued by a call to `self._controller.command`:

```
            data = self._controller.command(self._id, 'auto_conditioning_start')
            if data['response']['result']:
```

The `command` function in `controller.py` does not have the `@wake_up` decorator. Could that be the issue?

**Comment by alandtse** [comment]
@bart-roos I think you're right. I also think the wake_up decorators are on the wrong functions after update as those all deal with cached copies so waking up is not needed. I'll fix it and push it to staging.

**Comment by alandtse** [comment]
Beyond the wrapper, we need to fix all the other functions to handle the exceptions we'll run into if the wake_up fails.  Also, with a wrapper on update() we are hammering the wake up API for anything that is asleep because the wrapper itself isn't rate_limited. I am working on a solution but it may take a few days.

**Comment by alandtse** [comment]
@bart-roos I pushed a new version of staging that I think should address the issues with the decorator on the wrong functions. It also has some logging capability and will show up in HA logs if you enable debug for it:
```yaml
logger:
  logs:
    teslajsonpy.controller: debug
```
It currently will print out some debugging information whenever a wrapped function is called (update and command) and should give some idea if the vehicle woke up or not.

**Comment by bart-roos** [comment]
I am now testing the updated code. It seems that the `wake_up` is called, but the we either need to increase the delay between retries or increase the retry limit:

```
2019-02-11 15:02:16 DEBUG (Thread-21) [teslajsonpy.controller] Wakeup 21376123764343763: offline
2019-02-11 15:02:16 DEBUG (Thread-21) [teslajsonpy.controller] Result(0): False
2019-02-11 15:02:19 DEBUG (Thread-21) [teslajsonpy.controller] Wakeup 21376123764343763: offline
2019-02-11 15:02:19 DEBUG (Thread-21) [teslajsonpy.controller] Result(1): False
2019-02-11 15:02:21 DEBUG (Thread-7) [teslajsonpy.controller] Wrapper: f:<function Controller.update at 0x7feb790f1378>, result:None, args:(<teslajsonpy.controller.Controller object at 0x7feb790d48d0>, 21376123764343763), kwargs:{'wake_if_asleep': False},                                inst:<teslajsonpy.controller.Controller object at 0x7feb790d48d0>, vehicle_id:21376123764343763, __cars_online:{21376123764343763: False}
2019-02-11 15:02:21 DEBUG (Thread-21) [teslajsonpy.controller] Wakeup 21376123764343763: offline
2019-02-11 15:02:21 DEBUG (Thread-21) [teslajsonpy.controller] Result(2): False
2019-02-11 15:02:24 DEBUG (Thread-21) [teslajsonpy.controller] Wakeup 21376123764343763: offline
2019-02-11 15:02:24 DEBUG (Thread-21) [teslajsonpy.controller] Result(3): False
2019-02-11 15:02:27 DEBUG (Thread-21) [teslajsonpy.controller] Wakeup 21376123764343763: offline
2019-02-11 15:02:27 DEBUG (Thread-21) [teslajsonpy.controller] Result(4): False
2019-02-11 15:02:29 DEBUG (Thread-21) [teslajsonpy.controller] Wakeup 21376123764343763: offline
2019-02-11 15:02:29 DEBUG (Thread-21) [teslajsonpy.controller] Result(5): False
2019-02-11 15:02:30 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 241, in _step
    result = coro.throw(exc)
  File "/opt/homeassistant/lib/python3.5/site-packages/homeassistant/helpers/service.py", line 289, in _handle_service_platform_call
    await getattr(entity, func)(**data)
  File "/usr/lib/python3.5/asyncio/futures.py", line 380, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 304, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 293, in result
    raise self._exception
  File "/usr/lib/python3.5/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/homeassistant/lib/python3.5/site-packages/homeassistant/components/climate/tesla.py", line 98, in set_operation_mode
    self.tesla_device.set_status(False)
  File "/opt/homeassistant/lib/python3.5/site-packages/teslajsonpy/Climate.py", line 79, in set_status
    wake_if_asleep=True)
  File "/opt/homeassistant/lib/python3.5/site-packages/teslajsonpy/controller.py", line 105, in wrapped
    raise RetryLimitError
teslajsonpy.Exceptions.RetryLimitError
```

Just a few seconds later the log shows that the car has woken up:

```
2019-02-11 15:02:38 DEBUG (Thread-14) [teslajsonpy.controller] Wrapper: f:<function Controller.update at 0x7feb790f1378>, result:None, args:(<teslajsonpy.controller.Controller object at 0x7feb790d48d0>, 21376123764343763), kwargs:{'wake_if_asleep': False},                                inst:<teslajsonpy.controller.Controller object at 0x7feb790d48d0>, vehicle_id:21376123764343763, __cars_online:{21376123764343763: True}
```

This corresponds with what I see when using the official Tesla app. Sometimes it can take over 30 seconds for the car to wake up. I will try increasing the time between retries manually in the code to determine which value works reliably. By increasing the time, we prevent making too much calls to the Tesla API. I am not sure if that would be a problem at all, though.

**Comment by alandtse** [comment]
I'm thinking doing 5 retries with exponentially increasing delays may work. So retries at 2, 4, 8, 16 seconds gives a total of 30 seconds to wake up. If we're seeing consistently more than 30 seconds, we can start the first retry at 4 seconds delay. That means the final retry is at 1 minute.

**Comment by zabuldon** [comment]
@alandtse does make sense to display something to user in HA side?

**Comment by alandtse** [comment]
I'd prefer to leave some debugging in since you need to explicitly enable the debugging mode in a default HA configuration. This particular call could be commented out; I left something so @bart-roos could test it since I wasn't seeing it consistently.

**Comment by alandtse** [comment]
@bart-roos Any luck with the latest build? My internal testing seems to show it works, but would like confirmation before pushing a new version out on pypi.

**Comment by bart-roos** [comment]
The waking up seems to work after adjusting the retry settings, but turning on/off the HVAC from HA still does not work reliably. I would like to do some additional debugging, but didn't had the time yet to do so. I will be able to do this this or next week. I'll keep you posted on the progress.

**Comment by mlowijs** [comment]
With which vehicle/software version have you gotten the wake up to be reliable? My Model S with 2018.50.7 still does not want to wake up through the API (plugged in and in WiFi range).

**Comment by zabuldon** [comment]
@mlowijs did you use staging branch?

**Comment by bart-roos** [comment]
The wake-up procedure itself seems to work fine with the improved timing. There does seem to be another issue. What I see happening now is that the `_car_online` status is True, which skips the wake-up procedure. However, the Tesla API reports a `could_not_wake_buses` error. I discovered this issue by adding some extra debugging to the code to log the actual requests and responses to the Tesla API:

```
2019-02-19 22:21:08 DEBUG (Thread-5) [teslajsonpy.controller] Wrapper: f:<function Controller.update at 0x7f1824d46400>, result:None, args:(<teslajsonpy.controller.Controller object at 0x7f1834025400>, 13213213213123123), kwargs:{'wake_if_asleep': False}, inst:<teslajsonpy.controller.Controller object at 0x7f1834025400>, vehicle_id:13213213213123123, _car_online:{13213213213123123: True}
2019-02-19 22:21:09 DEBUG (Thread-2) [teslajsonpy.controller] get vehicles/13213213213123123/command/auto_conditioning_start
2019-02-19 22:21:09 DEBUG (Thread-2) [teslajsonpy.controller] {}
2019-02-19 22:21:11 DEBUG (Thread-2) [teslajsonpy.connection] {'response': {'reason': 'could_not_wake_buses', 'result': False}}
2019-02-19 22:21:11 DEBUG (Thread-2) [teslajsonpy.controller] Wrapper: f:<function Controller.update at 0x7f1824d46400>, result:None, args:(<teslajsonpy.controller.Controller object at 0x7f1834025400>, 13213213213123123), kwargs:{'wake_if_asleep': False}, inst:<teslajsonpy.controller.Controller object at 0x7f1834025400>, vehicle_id:13213213213123123, _car_online:{13213213213123123: True}
```

**Comment by alandtse** [comment]
@bart-roos For the always on problem, one thing that may be happening is the constant polling may not allow the Tesla to go to sleep. That is one issue we've been trying to address by disabling the wake_if_asleep but I do think we may not have a good mechanism in place yet.

On the "could_not_wake_buses" reason, we may be able to force the online state to off when we receive that and reschedule a retry. I'll look into it.

**Comment by bart-roos** [comment]
I am now running some tests with the latest version. Looking good so far :)

Another thing I noticed in my logging is that a call to get_vehicles is being made every second or so when the car is offline. This seems to be caused by the call in the wake_up decorator:

```
            if (cur_time - last_update > self.update_interval):
                cars = self.get_vehicles()
```

Wouldn't it be better to update the last_update time also if the car is offline? In that case the next try would happen only after another update_interval period (300 seconds in my config).

**Comment by bart-roos** [comment]
Just got a `could_not_wake_buses` again, but the retry was not triggered at all. The call just failed. I think there is an issue with this if statement:

```
            if (result is not None and
     --->       (result or
                 (('result' in result and not result['result']) or
                  ('reason' in result and result['reason'] !=
                    'could_not_wake_buses')))):
                return result
```

Does the second part of the if statement always evaluate to True as long as the result is not empty? If that is the case, the rest of the statement isn't relevant anymore because of the `or` operator.

**Comment by alandtse** [comment]
@bart-roos I think you're right on both issues. For the if statement, that line should be an "and" instead of an "or."

**Comment by bart-roos** [comment]
Now testing the updated version. For some reason, I stil see way to many calls to the  Tesla API being made:

```
2019-02-22 15:15:40 DEBUG (Thread-14) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:15:41 DEBUG (Thread-16) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:15:42 DEBUG (Thread-7) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:15:43 DEBUG (Thread-7) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:15:43 DEBUG (Thread-7) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:00 DEBUG (Thread-3) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:00 DEBUG (Thread-3) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:16:00 DEBUG (Thread-3) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:05 DEBUG (Thread-21) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:05 DEBUG (Thread-21) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:16:05 DEBUG (Thread-4) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:06 DEBUG (Thread-4) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:16:06 DEBUG (Thread-21) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:07 DEBUG (Thread-4) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:08 DEBUG (Thread-9) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:09 DEBUG (Thread-9) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:16:09 DEBUG (Thread-21) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:10 DEBUG (Thread-21) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:16:10 DEBUG (Thread-7) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:10 DEBUG (Thread-7) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
2019-02-22 15:16:10 DEBUG (Thread-2) [teslajsonpy.connection] /api/1/vehicles/21382736132137618/data
2019-02-22 15:16:11 DEBUG (Thread-2) [teslajsonpy.controller] Wrapped <function Controller.update at 0x7fac4c27d510> fails with False
Additional info: args:(<teslajsonpy.controller.Controller object at 0x7fac4c2516d8>, 21382736132137618), kwargs:{'wake_if_asleep': False}, vehicle_id:21382736132137618, _car_online:{21382736132137618: True}
```

These calls are being logged by an additional debugging statement I added to `connection.py`.

**Comment by alandtse** [comment]
Let's break the connection throttling issuing to a separate issue (#28) so we don't clutter this thread and just discuss the wake_up issue.

**Comment by alandtse** [comment]
@bart-roos How we looking on this particular issue? I think we have enough improvement we can probably resolve the thread on HA since that's been opened for months.

**Comment by bart-roos** [comment]
@alandtse I haven't done any further testing as I disabled the Tesla component in HA due to the throttling issue (#28). The large amount of requests seemed to cause connectivity issues between the official Tesla app and the car. I had issues with the car being completely unreachable by the app which resolved after I disabled the HA component.

**Comment by alandtse** [comment]
@bart-roos Ok thanks for confirming. That makes sense. We'll have to resolve it; I had not realized it was a new issue as opposed to an existing one.

**Comment by alandtse** [comment]
@bart-roos I refactored the code which should correctly address the wake_up_buses issue now. Please pull the latest staging version.  
`pip3 install --upgrade git+https://github.com/zabuldon/teslajsonpy.git@staging`

This may potentially address the throttle issue as I moved the retry code lower in the stack so it's not directly in the controller update which may have been causing the throttle issues for you (#28).  Can you please respond here if it resolves **the wake_up issues**?

**Comment by alandtse** [comment]
For the record, `could_not_wake_buses` appears to require a different solution then the original wake_up error. This is because that `could_not_wake_buses` error can happen even if the car is responding and is awake.  I've tweaked the wake_up function so the response to normal wake_up errors is to retry five wake up calls.  For `could_not_wake_buses`, it retries the same command up to five times.  

Right now it's separated into two phases so it's possible there will be a total of 10 total calls which will take 2 minutes based on the delay algorithm I posted earlier.

It may make sense to break it into two separate wrapper functions in the future.

**Comment by bart-roos** [comment]
Everything seems to work great in the latest staging version as far as I can see. I will do some real-life testing over the coming days to double-check. Great work!

**Comment by alandtse** [comment]
Great. I'll push this to master in the next day or so as I'm also testing in my production. Hopefully we've finally resolved this.

**Comment by zabuldon** [comment]
@alantse, great work! Thanks! Feel free to merge.

**Comment by alandtse** [comment]
@zabuldon Done and pushed to pypi. I'll submit the HA bump.
